hypervisor/dm/vpci/pci_pt.c

```diff
@@ -240,117 +178,56 @@ static void vdev_pt_deny_io_vbar(struct pci_vdev *vdev, uint32_t idx)
 	/* For SOS, all port IO access is allowed by default, so skip SOS here */
 	if (!is_sos_vm(vdev->vpci->vm)) {
 		struct pci_bar *vbar = &vdev->bar[idx];
-		if (vdev->bar_base_mapped[idx] != 0UL) {
-			deny_guest_pio_access(vdev->vpci->vm, (uint16_t)(vdev->bar_base_mapped[idx]),
-				(uint32_t)(vbar->size));
-			vdev->bar_base_mapped[idx] = 0UL;
+		if (vbar->base != 0UL) {
+			deny_guest_pio_access(vdev->vpci->vm, (uint16_t)(vbar->base), (uint32_t)(vbar->size));
+			vbar->base = 0UL;
 		}
 
 	}
 }
 
-/**
- * @brief Set the base address portion of the vbar base address register (32-bit)
- * base: bar value with flags portion masked off
- * @pre vbar != NULL
- */
-static void set_vbar_base(struct pci_bar *vbar, uint32_t base)
-{
-	union pci_bar_reg bar_reg;
-
-	bar_reg.value = base;
-
-	if (vbar->is_64bit_high) {
-		/* Upper 32-bit of a 64-bit bar does not have the flags portion */
-		vbar->reg.value = bar_reg.value;
-	} else if (vbar->reg.bits.io.is_io == 1U) {
-		/* IO bar, BITS 31-2 = base address, 4-byte aligned */
-		vbar->reg.bits.io.base = bar_reg.bits.io.base;
-	} else {
-		/* MMIO bar, BITS 31-4 = base address, 16-byte aligned */
-		vbar->reg.bits.mem.base = bar_reg.bits.mem.base;
-	}
-}
-
 /**
  * @pre vdev != NULL
  */
-static void vdev_pt_write_vbar(struct pci_vdev *vdev, uint32_t offset, uint32_t val)
+void vdev_pt_write_vbar(struct pci_vdev *vdev, uint32_t idx, uint32_t val)
 {
-	uint32_t idx;
-	uint64_t base;
-	bool bar_update_normal;
-	struct pci_bar *vbar;
-
-	base = 0UL;
-	idx = (offset - pci_bar_offset(0U)) >> 2U;
-	bar_update_normal = (val != (uint32_t)~0U);
-
-	vbar = &vdev->bar[idx];
+	bool update_bar = false;
+	uint32_t update_idx = idx;
+	uint32_t offset = pci_bar_offset(idx);
+	struct pci_bar *vbar = &vdev->bar[idx];
 
-	if (vbar->is_64bit_high) {
-		if (idx > 0U) {
-			uint32_t prev_idx = idx - 1U;
-
-			vdev_pt_unmap_mem_vbar(vdev, prev_idx);
-			base = git_size_masked_bar_base(vdev->bar[prev_idx].size, ((uint64_t)val) << 32U) >> 32U;
-			set_vbar_base(vbar, (uint32_t)base);
-
-			if (bar_update_normal) {
-				vdev_pt_map_mem_vbar(vdev, prev_idx);
-			}
+	switch (vbar->type) {
+	case PCIBAR_IO_SPACE:
+		vdev_pt_deny_io_vbar(vdev, update_idx);
```

Comment:
deny_old_bar, set the new bar, and then allow_new_bar...
I commented many times, but was ignored many times